### ISSUE TITLE
Object#download_file Add Progress Callback

### ISSUE THREAD

**Issue by foxtrod** [aws/aws-sdk-ruby#2901]
### Describe the feature

Progress Callback for Object#download_file

Introduce a built-in mechanism within the Object#download_file that provides real-time feedback on the download progress just like in a [Object#upload_file](https://docs.aws.amazon.com/sdk-for-ruby/v3/api/Aws/S3/Object.html#upload_file-instance_method)

### Use Case

Seeing is Believing.

Currently, to get any sense of progress, I was left either diving deep into the AWS client logger or constantly pinging the file size, I've picked the second option and wondering how users of the gem are handling the progress (let me know in the comments). The Object#upload_file method has this awesome progress feedback, so why shouldn't Object#download_file have the same?

### Acknowledgements

- [X] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### SDK version used

3.180.3

### Environment details (OS name and version, etc.)

MacOs Ventura 13.3

**Comment by foxtrod** [comment]
Would be glad if i missed existing functionality that covers download progress, but if someone wonders I'm using next workaround:

```
download_thread = Thread.new do
  s3_object.download_file(destination, { thread_count: threads_count, mode: "get_range", chunk_size: chunk_size })
end

progress_thread = Thread.new do
  begin
    while download_thread.alive? # ping file size until download processing
      current_size = File.exist?(destination) ? File.size(destination) : 0
      progress_percentage = (current_size.to_f / expected_size) * 100

      puts "Downloaded: #{current_size} / #{expected_size} (#{progress_percentage.round(2)}%)"

      break if current_size >= expected_size

      sleep 1 # wait for a second before checking again
    end
  rescue => e
    logger.error(event: :progress_thread, error: e)
  end
end

download_thread.join 
progress_thread.join
```

**Comment by jterapin** [comment]
Resolved in #2902.
